lib/probes/http.js

```diff
@@ -128,12 +135,37 @@ function patchClient (module, options, protocol) {
       // Do request. args will be different than the caller's original args
       // if no options argument was supplied because we need to add an
       // x-trace header and that requires an options argument.
-      ret = fn(...args)
+      ret = fn(...args);
 
       try {
 
-        // Report socket errors
-        ret.prependListener('error', error => span.error(error));
+        // make sure the span is exited on a socket error.
+        ret.prependListener('error', error => {
+          span.error(error);
+          span.exit();
+        });
+        // it doesn't appear that this needs to listen for the 'abort' event.
+        // in all cases i've been able to reproduce either the 'end' or the
+        // 'error' event is called.
+
+        // Handle Websocket upgrade
+        ret.prependListener('upgrade', (res, socket, head) => {
+          // and exit the span immediately. it's not possible to know how long
+          // the websocket session will exist. documentation-wise we'll need
+          // to note that we only instrument the upgrade and that any
+          // instrumentation of transactions over an open web socket must use
+          // the custom instrumentation API. also note that if the server
+          // provides a bad sec-websocket-accept header value then the client
+          // can (and should) reject the connection but we won't know that here
+          // because the 'upgrade' listener is expected to perform that check.
+          //
+          // the upgrade event is only emitted if the status code is 101 so it
+          // be hardcoded. but the message is whatever the server sets it to.
```

Comment:
what i meant was that i could hardcode the status as 101 because node will not emit an upgrade event unless the status is 101. i decided not to hardcode it on the wild chance that node might emit an upgrade event for some other status code.